src/com/xilinx/rapidwright/rwroute/RWRoute.java

```diff
@@ -1484,20 +1495,59 @@ protected void checkPIPsUsage() {
             System.out.println("\nINFO: No PIP overlaps\n");
     }
 
+    /**
+     * Class encapsulating all state necessary to route the included connection
+     */
+    protected static class ConnectionState {
+        protected final Connection connection;
+        protected final int sequence;
+        protected final float rnodeCostWeight;
+        protected final float shareWeight;
+        protected final float rnodeWLWeight;
+        protected final float estWlWeight;
+        protected final float dlyWeight;
+        protected final float estDlyWeight;
+        protected final PriorityQueue<RouteNode> queue;
+        protected final List<RouteNode> targets;
+
+        ConnectionState(Connection connection, int sequence, float rnodeCostWeight, float shareWeight,
+                        float rnodeWLWeight, float estWlWeight, float dlyWeight, float estDlyWeight,
+                        PriorityQueue<RouteNode> queue) {
+            this.connection = connection;
+            this.sequence = sequence;
+            this.rnodeCostWeight = rnodeCostWeight;
+            this.shareWeight = shareWeight;
+            this.rnodeWLWeight = rnodeWLWeight;
+            this.estWlWeight = estWlWeight;
+            this.dlyWeight = dlyWeight;
+            this.estDlyWeight = estDlyWeight;
+            this.queue = queue;
+            targets = new ArrayList<>();
```

Comment:
Seems like adding a new object for each connection in a design could impact memory--even if it is just accumulating a bunch of parameters that were floating around previously.  It does look like the `targets` list is new.  Would it make sense to add them directly to the `Connection` object itself to avoid some of the overhead?

Reply:
This is a new object that is only live when this `Connection` is being routed, which for the single-threaded case is at most one object (mostly removed from the function stack). Putting it into the `Connection` object means paying this cost for all connections regardless of whether they are currently being routed.

For the single-threaded case again, the `targets` list is moved from `RouteNodeGraph` into this new short-lived `ConnectionState` object.

I agree that creating and destroying `ConnectionState` now does create a more garbage though.

Reply:
Ok, makes sense so no real impact to memory footprint.  

You could consider a `static` `ThreadLocal` instance if you are worried about garbage collector thrashing, but I wonder if the latest JVMs compensate for this pattern.

Reply:
My first approach was to use `ThreadLocal`, but I'm assuming that a `ThreadLocal.get()` incurs a non-trivial overhead such that it doesn't make sense to call it inside the inner loop (e.g. the `push()` method called for each node that is put on the queue) if you can pass it as a function parameter instead.